app/gui2/src/components/GraphEditor/GraphNodeComment.vue

```diff
@@ -16,101 +15,148 @@ const emit = defineEmits<{
 }>()
 
-const paragraphs = computed(() => props.modelValue.split('\n\n'))
+const text = computed({
+  get: () => rawTextToCooked(props.modelValue),
+  set: (value) => emit('update:modelValue', cookedTextToRaw(value)),
+})
 
-const contentElement = ref<HTMLElement>()
+const commentRoot = ref<HTMLElement>()
 const editor = ref<EditorViewType>()
 
-const interaction = injectInteractionHandler()
-interaction.setWhen(() => editor.value != null, {
-  cancel() {
-    finishEdit()
-  },
-  click(e: Event) {
-    if (e.target instanceof Element && !contentElement.value?.contains(e.target)) finishEdit()
+const interactions = injectInteractionHandler()
+const editInteraction = {
+  cancel: () => finishEdit(),
+  click: (e: Event) => {
+    if (e.target instanceof Element && !commentRoot.value?.contains(e.target)) finishEdit()
     return false
   },
-})
-
-const handleClick = defineKeybinds(`comment-${random.uint53()}`, {
-  startEdit: ['Mod+PointerMain'],
-}).handler({ startEdit })
+}
+interactions.setWhen(() => props.editing, editInteraction)
+onUnmounted(() => interactions.end(editInteraction))
 
 function startEdit() {
-  if (editor.value) {
-    editor.value.focus()
-  } else {
-    const editorView = new EditorView()
-    editorView.setState(EditorState.create({ extensions: [minimalSetup] }))
-    contentElement.value!.prepend(editorView.dom)
-    editor.value = editorView
-    setTimeout(() => editorView.focus())
-  }
   if (!props.editing) emit('update:editing', true)
 }
 
 function finishEdit() {
-  if (editor.value) {
-    if (editor.value.state.doc.toString() !== props.modelValue)
-      emit('update:modelValue', editor.value.state.doc.toString())
-    editor.value.dom.remove()
-    editor.value = undefined
+  if (props.editing) {
+    if (editor.value) {
+      const viewText = editor.value.state.doc.toString()
+      if (viewText !== text.value) text.value = viewText
+    }
+    emit('update:editing', false)
   }
-  if (props.editing) emit('update:editing', false)
+}
+
+function insertTextAtCursor(insert: string) {
+  if (!editor.value) return
+  const range = editor.value.state.selection.ranges[0] ?? { from: 0, to: 0 }
+  editor.value.dispatch({
+    changes: {
+      from: range.from,
+      to: range.to,
+      insert,
+    },
+    selection: { anchor: range.from + insert.length },
+  })
+}
+
+function handleEnter(event: KeyboardEvent) {
+  if (event.shiftKey) insertTextAtCursor('\n')
+  else finishEdit()
 }
 
 watchEffect(() => {
-  const text = props.modelValue
   if (!editor.value) return
   const viewText = editor.value.state.doc.toString()
   editor.value.dispatch({
-    changes: textChangeToEdits(viewText, text).map(textEditToChangeSpec),
+    changes: textChangeToEdits(viewText, text.value).map(textEditToChangeSpec),
   })
 })
 
 watchEffect(() => {
-  if (contentElement.value && props.editing && !editor.value) startEdit()
+  if (!editor.value) return
+  if (props.editing) editor.value.focus()
+  else editor.value.contentDOM.blur()
```

Comment:
This looks little strange as a `watchEffect`. Could this focus/blur be moved to the interaction methods? Or is there any reason for it being here in particular?

Reply:
Watching the prop ensures that focus/blur is updated when we begin/end editing for any reason, e.g:
- If the edit-comment menu icon is clicked, that sets the prop, so the comment is focused when editing begins
- There are various ways editing can be ended, some of them will set a new focus inherently but some don't; if there were any case where we missed unfocusing the comment, a weird state would result in which the app thinks the comment is not being edited but the user still sees a cursor.